# Advent of Code 2015, Dyalog APL edition

In [1]:
⎕IO←1
]box on -style=max -trains=tree -fns=on ⍝ Pass all output through DISPLAY

Rebuilding user command cache... done

┌→─────────────────────────────────────┐
│Was ON -style=min -trains=tree -fns=on│
└──────────────────────────────────────┘

### Day 1: Not Quite Lisp
https://adventofcode.com/2015/day/1

In [2]:
DAY1←⊃⊃⎕NGET'data/2015/01.txt'1

In [3]:
Day1p1←{+/+⌿-@2⌷'()'∘.=⍵}     ⍝ Binary mask with positions for '(' in row 1 and ')' in row 2, 
                              ⍝ and then row 2 negated. Sum down, then across.

In [4]:
Day1p1 DAY1 ⍝ 138

138

In [5]:
Day1p2←{1⊃⍸{⍵<0}¨+\+⌿-@2⌷'()'∘.=⍵} ⍝ As before, but scan-sum across and create mask of negatives and pick first.

In [6]:
Day1p2 DAY1 ⍝ 1771

1771

### Day 2: I Was Told There Would Be No Math
https://adventofcode.com/2015/day/2

In [7]:
DAY2←2⊃¨'x'⎕VFI¨⊃⎕NGET 'data/2015/02.txt' 1  ⍝ Many lines. Pick out three ints from each, separated by 'x'

In [8]:
Day2p1←{(l w h)←⍵ ⋄ s[⊃⍋s]++/2×s←(l×w) (w×h) (h×l)}

In [9]:
+/Day2p1¨DAY2 ⍝ 1606483

1606483

In [10]:
Day2p2←{(ll ww hh)←2×⍵⋄c[⊃⍋c←(hh+ww) (ww+ll) (ll+hh)]+×/⍵}

In [11]:
+/Day2p2¨DAY2 ⍝ 3842356

3842356

### Day 3: Perfectly Spherical Houses in a Vacuum

https://adventofcode.com/2015/day/3

In [12]:
DAY3←⊃⊃⎕NGET 'data/2015/03.txt' 1

In [13]:
]dinput
Day3p1←{
    points←-@1 3⌷'^v<>'∘.=⍵                   ⍝ Matrix with 4 rows -y, y, -x, x
    ⍉(+\+⌿points[3 4;]),[0.5]+\+⌿points[1 2;] ⍝ Sum first along axis 1, then sum-scan along axis 2
}                                             ⍝ separately for x and y. Laminate x and y parts, transpose

In [14]:
1+≢∪ Day3p1 DAY3 ⍝ 2081

2081

In [15]:
Day3p2←{(1⊃p)⍪2⊃p←Day3p1¨{⍵⊢∘⊂⌸⍨2|⍳≢⍵}⍵}      ⍝ Unzip ⍵ so we can treat santa and robo's paths separately

In [16]:
≢∪ Day3p2 DAY3 ⍝ 2341

2341

### Day 4: The Ideal Stocking Stuffer
https://adventofcode.com/2015/day/4

In [27]:
Md5←{⎕SH 'md5 -q -s "',⍵,'"'}  ⍝ No Dyalog built-in MD5 function, sadly

In [28]:
Day4p1←{⍺←0⋄'00000'≡5↑Md5 ⍵,⍕⍺:⍺⋄(⍺+1)∇⍵}  ⍝ NOTE: WORKS, BUT INTRACTABLE WITHOUT NATIVE MD5

### Day 5: Doesn't He Have Intern-Elves For This?
https://adventofcode.com/2015/day/5

In [5]:
DAY5←⊃⎕NGET 'data/2015/05.txt' 1

In [8]:
]dinput
Day5p1←{
    3>+/+⌿'aeiou'∘.=⍵:0                  ⍝ Three or more vowels
    0≠+/'ab' 'cd' 'pq' 'xy'∊↓{⍵}⌺2⊢⍵:0   ⍝ Illegal pairs
    1≤+/=⌿⍵,[0.5](≢⍵)↑1↓⍵                ⍝ Repeated letters
}

In [9]:
+/Day5p1¨DAY5 ⍝ 238

238

In [10]:
]dinput
Day5p2←{
    0=≢'(..).*\1'⎕S'\1'⊢⍵:0              ⍝ Regexery feels like a defeat
    0≠+/(⊣/=⊢/){⍵}⌺3⊢⍵                   ⍝ Overlapping triplets, compare first and last columns
 }

In [11]:
+/Day5p2¨DAY5 ⍝ 69

69

### Day 6: Probably a Fire Hazard
https://adventofcode.com/2015/day/6

This time we'll go old-skool with a tradfn.

In [19]:
⎕IO←0       ⍝ The problem's coordinates are 0-based.
⎕CY 'dfns'  ⍝ iotag -- generalised iota -- alows us to create ranges from a to b

In [20]:
DAY6←⊃⎕NGET 'data/2015/06.txt' 1

In [21]:
RegexGroups←{⍺⎕S{⍵.(1↓Lengths↑¨Offsets↓¨⊂Block)} ⊢ ⍵}

In [24]:
]dinput
res←Day6p1 data;state;parsed;row;spec;points
state←1000 1000⍴0
parsed←'(on|off|toggle)\s(\d+),(\d+)[^\d]+(\d+),(\d+)' RegexGroups data
:For row :In parsed
    spec←⍎¨1↓row
    points←(spec[0] iotag spec[2]) ∘., (spec[1] iotag spec[3])
    :If (0⊃row)≡'on'
        state[points]←1
    :ElseIf (0⊃row)≡'off'
        state[points]←0
    :Else
        state[points]←~state[points]
    :EndIf
:EndFor
res←+/+⌿state

In [23]:
Day6p1 DAY6 ⍝ 569999

569999

Part 2 is more of the same, just with different updates: turn on means add 1, turn off means subtract 1, capped at 0, and toggle means add 2.

In [25]:
]dinput
res←Day6p2 data;state;parsed;row;op;spec;points
state←1000 1000⍴0
parsed←'(on|off|toggle)\s(\d+),(\d+)[^\d]+(\d+),(\d+)' RegexGroups data
:For row :In parsed
    spec←⍎¨1↓row
    points←(spec[0] iotag spec[2]) ∘., (spec[1] iotag spec[3])
    :If (0⊃row)≡'on'
        state[points]+←1
    :Elseif (0⊃row)≡'off'
        state[points]-←1
        state[points]⌈←0
    :Else
        state[points]+←2
    :EndIf
:EndFor
res←+/+⌿state

In [26]:
Day6p2 DAY6 ⍝ 17836115

17836115

### Day 7: Some Assembly Required
https://adventofcode.com/2015/day/7

A very interesting question. We have a set of "gates", specified like so:

```
af AND ah -> ai
NOT lk -> ll
hz RSHIFT 1 -> is
NOT go -> gp
du OR dt -> dv
```

So, for example, the gate "ai" has the value of the bitwise AND of the values of gates "af" and "ah". In other words, the gates are _functions_ with inputs and outputs, and the input data is already a full program. We can exploit this in our solution. Take

    af AND ah -> ai
  
and turn this into an APL function

    ai←{(af ⍬) AND (ah ⍬)}
    
assuming that we have the AND function defined. If we convert the input specifications to such APL functions, all we need to do is to run the "a" function to get our result. Elegant, but with a small problem: it doesn't work. 

The reason for this is that the functions can be expensive to compute, as may require many, many calls. However, as they never _change_, we can cache the values after the first invocation, and with this in place the solution is found very quickly. Instead of the above expression, we make

    ai←({⍵⊣⍵.c←¯1}⎕NS'') Cache {(af ⍬) AND (ah ⍬)}

with the Cache operator defined as 

    Cache←{¯1=⍺⍺.c:⍺⍺.c⊣⍺⍺.c←⍵⍵ ⍵⋄⍺⍺.c}
    
which takes a namespace holding the cached int value.

Health warning: there is no macro facility in APL, so instead we're abusing the hydrant (⍎) function to evaluate APL code represented as strings. In lisp lingo, this is 100% unhygenic and rather unsafe.

In [84]:
⎕IO←1

In [85]:
DAY7←⊃⎕NGET 'data/2015/07.txt' 1

In [86]:
⍝ Some helper functions
Bin←{(16⍴2)⊤⍵}
Dec←{2⊥⍵}
RegexGroups←{⍺⎕S{⍵.(1↓Lengths↑¨Offsets↓¨⊂Block)} ⊢ ⍵}
Cache←{¯1=⍺⍺.c:⍺⍺.c⊣⍺⍺.c←⍵⍵ ⍵⋄⍺⍺.c}

In [87]:
⍝ Gate functions as defined in the problem statement
AND←{Dec (Bin ⍺)∧(Bin ⍵)}
OR←{Dec (Bin ⍺)∨(Bin ⍵)}
NOT←{Dec ~(Bin ⍵)}
RSHIFT←{ba←Bin ⍺⋄Dec (-≢ba)↑(-⍵)↓ba}
LSHIFT←{ba←Bin ⍺⋄Dec (≢ba)↑⍵↓ba}

In [88]:
Arg←{1=1⊃⎕VFI ⍵:⍵ ⋄ ∊'(',⍵,' ⍬)'}

In [89]:
Parse←{'(\w*)\s*(AND|NOT|OR|RSHIFT|LSHIFT|)\s*(\w*)\s->\s(\w+)' RegexGroups ⍵}

In [90]:
]dinput
Compile←{ ⍝ Turn the parsed input to strings representing APL functions.
    (⊃⍵[1])≡'NOT':∊⍵[4],'←({⍵⊣⍵.c←¯1}⎕NS'''') Cache {NOT ',(Arg ⊃⍵[3]),'}'
    0=≢⊃⍵[2]:∊⍵[4],'←({⍵⊣⍵.c←¯1}⎕NS'''') Cache {',(Arg ⊃⍵[1]),'}'
    ∊⍵[4],'←({⍵⊣⍵.c←¯1}⎕NS'''') Cache {',(Arg ⊃⍵[1]),' ',⍵[2],' ',(Arg ⊃⍵[3]),'}'
}

In [91]:
⍎¨Compile¨Parse DAY7

In [93]:
a ⍬                   ⍝ Part1: what's the value of "a"? => 956

956

Part2: Reset, and then override "b" to the final value of "a" in Part 1. What's the value of "a" now?

In [96]:
⍎¨Compile¨Parse DAY7  ⍝ Reset by recompiling

In [97]:
b←{956}               ⍝ Override "b" -- no need for the Cache.

In [98]:
a ⍬                   ⍝ Part2: what's the value of "a"? =>40149

40149

In [99]:
)clear

clear ws

### Day 8: Matchsticks
https://adventofcode.com/2015/day/8

In [105]:
⎕IO←1

In [123]:
DAY8TEXT←⊃⎕NGET 'data/2015/08.txt' 1
DAY8BYTES←⊃{⍺⎕UCS¨⍨⊂'-\w+$'⎕R''⊢⍵}/2↑⎕NGET 'data/2015/08.txt' 1  ⍝ Raw bytes, split on line endings

In [134]:
Day8p1←{(≢⍵)-¯2+≢'\\\\'⎕R'\\'⊢s←'(\\x[a-f0-9]{2}|\\[\''"])'⎕R'⎕'⊢⍵}

In [135]:
+/Day8p1¨DAY8TEXT ⍝ 1350

1350

In [138]:
Day8p2←{(≢⍵)-⍨(2+2×+/bq)++/~bq←3≠34 92⍳⍵} ⍝ +2 for surrounding brackets, +2 for every " and \

In [137]:
+/Day8p2¨DAY8BYTES ⍝ 2085

2085

### Day 9: All in a Single Night
https://adventofcode.com/2015/day/9

In [295]:
⎕IO←1
⎕CY 'dfns' ⍝ pmat

In [296]:
DAY9←' '(≠⊆⊢)¨'( to | = )'⎕R' '⊃⎕NGET'data/2015/09.txt'1

In [297]:
]dinput
Day9←{
     weights←w,w←⍎¨⊢/↑⍵                 ⍝ Two copies of the weights, as A->B == B->A 
     edges←(⊢⍪⌽)28 2⍴(∪⍳⊢),(↑⍵)[;1 2]   ⍝ Swap strings for numeric index, and add reverse direction
     adj←weights@(↓edges)⊢0⍴⍨2⍴8        ⍝ Build adjacency matrix
     (⌊/,⌈/)+/adj[↓{∊⍵}⌺1 2⊢pmat 8]     ⍝ Generate length 8 permuations, grab the path length, then min-max.
 }

In [298]:
Day9 DAY9 ⍝ p1: 251 p2: 898

251 898

### Day 10: Elves Look, Elves Say
https://adventofcode.com/2015/day/10

In [300]:
DAY10←'3113322113'

In [301]:
Day10←{⍵=0:≢⍺⋄(∊(⍕∘≢,1∘↑)¨({⍵⊂⍨1,2≠/⍵} ⍺))∇⍵-1} ⍝ For each group of equal elements, generate count+element

In [303]:
DAY10 Day10 40  ⍝ 329356

329356

In [304]:
DAY10 Day10 50  ⍝ 4666278

4666278

### Day 11: Corporate Policy
https://adventofcode.com/2015/day/11

In [330]:
⎕IO←1
AB←↓⍉↑(⊂,⊂)⎕A      ⍝ Repeated pairs: aa bb cc ...
ABC←¯1↓1↓↓{⍵}⌺3⊢⎕A ⍝ Length-3 straights: abc bcd cde ...

In [325]:
]dinput
Next←{
    ⍝ Generate the "next" string by increasing the rightmost letter
    ⍝ and carrying on leftwards if we roll over Z->A.
    abc←⎕A,'A'
    ⍵ {
        l←abc[1+abc⍳⍺[⍵]]
        l≢'A':l@⍵⊢⍺
        (l@⍵⊢⍺)∇⍵-1
    } ≢⍵
}

In [326]:
]dinput
Valid←{
    ∨/'OIL'∊⍵:0             ⍝ Rule 2
    0=+/ABC∊¯1↓1↓↓{⍵}⌺3⊢⍵:0 ⍝ Rule 1
    (+/AB∊{⊂⍵}⌺2⊢⍵)≥2       ⍝ Rule 3
}

In [327]:
]dinput
Day11←{
    next←Next ⍵
    Valid next:next
    ∇next
}

In [328]:
Day11 'HEPXCRRQ'  ⍝ HEPXXYZZ

HEPXXYZZ

In [329]:
Day11 'HEPXXYZZ'  ⍝ HEQAABCC

HEQAABCC

### Day 12: JSAbacusFramework.io
https://adventofcode.com/2015/day/12

In [71]:
]dinput
ValidItems←{
    values←{0=≢⍵.(⎕NL ¯2):⍬⋄⍵.(⍎¨⎕NL ¯2)} ⍵ 
    (⊂'red')∊values:,0
    0=≢⍵.(⎕NL ¯9):values
    values,∊∇¨⍵.(⍎¨⎕NL ¯9)
}

In [69]:
]dinput
Day12p2←{
    {(1=2|⎕DR)⍵}⍵:⍵                        ⍝ Are we a number?
    (⍕≡⊢)⍵:0                               ⍝ Are we a string?
    {(326=⎕DR⍵)∧(0=≡)⍵} ⍵:+/∇ValidItems ⍵  ⍝ Are we an object?
    +/∊∇¨⍵                                 ⍝ We're a list
}

In [72]:
{+/(3=⊢/⍵)/⍵[;3]} (⎕JSON⍠'M')⊃⊃⎕NGET'data/2015/12.txt'1 ⍝ Part 1: just sum all values...156366
Day12p2 ⎕JSON⊃⊃⎕NGET'data/2015/12.txt'1                 ⍝ Part 2: 96852

156366

96852

### Day 13: Knights of the Dinner Table
https://adventofcode.com/2015/day/13

This is very similar to day 9, and we can broadly use the same technique.

In [444]:
⎕IO←1
RegexGroups←{⍺⎕S{⍵.(1↓Lengths↑¨Offsets↓¨⊂Block)} ⊢ ⍵}
DAY13← ↑'(.+) would (gain|lose) (\d+).+to ([^.]+)' RegexGroups ⊃⎕NGET'data/2015/13.txt'1

In [445]:
EDGES←⍉(⊣/DAY13),[0.5]⊢/DAY13
WEIGHTS←-@(((⊂'lose')≡¨DAY13[;2])/⍳≢DAY13[;2]) ⊢ ⍎¨DAY13[;3]

In [446]:
e←(⍴EDGES)⍴(∪⍳⊢),EDGES    ⍝ Convert strings to indexes
adj←WEIGHTS@(↓e)⊢0⍴⍨2⍴8   ⍝ Happiness matrix

In [447]:
adj                       ⍝ Note: asymmetric!

┌→──────────────────────────────┐
↓ 0 2 26 ¯82 ¯75 42 38 39│
│ 40 0 ¯61 ¯15 63 41 30 87│
│¯35 ¯99 0 ¯51 95 90 ¯16 94│
│ 36 ¯18 ¯65 0 ¯18 ¯22 2 42│
│¯65 24 100 51 0 21 55 ¯44│
│¯48 91 8 ¯66 97 0 ¯9 ¯92│
│¯44 ¯25 17 92 ¯92 18 0 97│
│ 92 ¯96 ¯51 ¯81 31 ¯73 ¯89 0│
└~──────────────────────────────┘

Unlike Day 9, this graph is directional, so the weight from A to B is not the same as the weight from B to A. So we need to generate all permutations of length 8, but then ensure that we also have the reverse directions -- and also ensure that the end is tied back to the beginning.

The expression `p,⊣/p←pmat 8` copies the first column to the end to ensure the circle is connected, and `m,1↓[2]⌽m` means that we add in all the reverse connections.

In [450]:
⌈/+/adj[↓{∊⍵}⌺1 2⊢m,1↓[2]⌽m←p,⊣/p←pmat 8] ⍝ Part 1: 733

733

For part 2 we simply don't connect the circle -- this has the same effect as inserting a weight-0 node.

In [451]:
⌈/+/adj[↓{∊⍵}⌺1 2⊢m,1↓[2]⌽m←pmat 8] ⍝ Part 2: 725

725

### Day 14: Reindeer Olympics
https://adventofcode.com/2015/day/14

In [475]:
⎕IO←1
DAY14←⍎¨'\d+'⎕S'&'⊃⎕NGET'data/2015/14.txt'1

In [476]:
DATA←(≢DAY14)3⍴DAY14

In [477]:
⍝ Speed, Travel-time, Rest-time, State, Counter, Distance travelled, Score
LB←(((DATA,1),(DATA[;2])),0),0  

In [481]:
]dinput
Day14p1←{
    lb←⍵
    lb[;5]-←1                            ⍝ Decrease the counter
    lb[;6]+←lb[;1]×1=lb[;4]              ⍝ Move any movers; resters unchanged
    lb[;7]+←(⌈/a)=a←lb[;6]               ⍝ Add 1 to the leaders col for part 2.
    ↑{0≠⍵[5]:⍵⋄⍵[3-s[4]]@5⊢s←~@4⊢⍵}¨↓lb  ⍝ Perform state transitions where counter is 0
}

In [484]:
RACE←Day14p1⍣2503⊢LB                     ⍝ Run for 2503 iterations
⌈/RACE[;6]                               ⍝ Part 1 is max distance: 2655
⌈/RACE[;7]                               ⍝ Part 2 is max score: 1059

2655

1059

### Day 15: Science for Hungry People
https://adventofcode.com/2015/day/15

In [62]:
⎕IO←1
DAY15←(5÷⍨≢d)5⍴d←⍎¨'-?\d+'⎕S'&'⊃⎕NGET'data/2015/15.txt'1 ⍝ The numbers, wrestled into matrix of 5-element rows

In [63]:
CAL←⊢/DAY15      ⍝ Keep calorie data in a vector
INGR←¯1↓[2]DAY15 ⍝ Just the ingredients

In [64]:
Cmb←{(n t)←⍵⋄A←⊃∘.+/(n-1)⍴⊂⍳t⋄⊃,/{⍵,¨⍸(t-⍵)=A}¨⍳t} ⍝ Cmb N M -- All N-length vectors summing to M, N[i]∊⍳M

In [65]:
Day15p1←{ingr←⍵⋄⌈/{×/0∘⌈¨⍵+.×ingr}¨Cmb 4 100} ⍝ Multiply with cost matrix, multiply again, removing negatives

In [66]:
Day15p2←{ingr←⍵⋄cal←⍺⋄⌈/{500=+/cal×⍵:×/0∘⌈¨⍵+.×ingr⋄0}¨Cmb 4 100} ⍝ As above, but exactly 500 cals.

In [67]:
Day15p1 INGR ⍝ 222870
CAL Day15p2 INGR ⍝ 117936

222870

117936

### Day 16: Aunt Sue
https://adventofcode.com/2015/day/16

In [78]:
⎕IO←1
DAY16←⊃⎕NGET'data/2015/16.txt'1
TARGET←0 2 7 3 5 1 3 2 3 0
ITEMS←'aki' 'car' 'cat' 'chi' 'gol' 'per' 'pom' 'sam' 'tre' 'viz'

In [79]:
]dinput
ToRow←{
    m←↑{(⊂3↑1↓1⊃⍵),⊂⍎1↓2⊃⍵}¨':'(≠⊆⊢)¨⍵ ⍝ Split on ':', pick first 3 letters and convert the numbers
    r←10⍴0
    r[ITEMS⍳⊣/m]←⊢/m                   ⍝ Fill in the known quantities
    r
}

In [80]:
Parse←{ToRow¨','(≠⊆⊢)¨'Sue \d+:'⎕R''⊢⍵}

In [81]:
]dinput
Day16p1←{
    where←⍸0≠⍵
    ⍺[where]≡⍵[where]
}

In [98]:
]dinput
Day16p2←{
    target←⍺
    sue←⍵

    ∧/{
        (⍵=3)∨⍵=9:sue[⍵]>target[⍵]
        (⍵=5)∨⍵=7:sue[⍵]<target[⍵]
        sue[⍵]=target[⍵]
    }¨⍸0≠⍵
}

In [99]:
⍸TARGET∘Day16p1¨Parse DAY16 ⍝ Part 1: 40
⍸TARGET∘Day16p2¨Parse DAY16 ⍝ Part 2: 241 TODO: should not also return 225...

┌→─┐
│40│
└~─┘

┌→──────┐
│225 241│
└~──────┘

### Day 17: No Such Thing as Too Much
https://adventofcode.com/2015/day/17

Count the number of subsets of buckets that can hold exactly 150 units. This can be cast as a count-up from 1 to a "bucket-count"-bit binary number of all 1s. A 1 in a given place means that the corresponding bucket is picked. For example, if we have the following five buckets:

    50 44 11 49 42
    
we need to count from 1 to 31 (11111). If we take for example the seventh (00111) selection we get

    50 44 11 49 42
     0  0  1  1  1  => 11 + 49 + 42 = 102
     
We can translate this approach directly to APL.

In [102]:
DAY17←50 44 11 49 42 46 18 32 26 40 21 7 18 43 10 47 36 24 22 40 ⍝ Container volumes
Bin←{(20⍴2)⊤⍵}
Dec←{2⊥⍵}

In [103]:
+/{150=+/DAY17/⍨Bin ⍵}¨⍳Dec 20⍴1 ⍝ Part 1: 654

654

For part 2 we need to find the fewest number of buckets that can hold 150 units, and then count the number of distinct such combinations. We know there are 654 solutions in total -- let's capture them, rather than just counting them:

In [104]:
VALID←(0≠v)/v←{150=+/DAY17/⍨Bin ⍵:⍵⋄0}¨⍳Dec 20⍴1 ⍝ Return the sequence number if sum is 150, otherwise 0

Next step is to find the fewest number of buckets:

In [115]:
⎕←MIN←⌊/≢¨BUCKETS←{DAY17/⍨Bin ⍵}¨VALID

4

and then count the solutions using the minimum number of buckets, which is our answer:

In [117]:
+/MIN=≢¨BUCKETS ⍝ 57

57

### Day 18: Like a GIF For Your Yard
https://adventofcode.com/2015/day/18

This is an implementation of [Conway's Game of Life](https://en.wikipedia.org/wiki/Conway%27s_Game_of_Life). It's a bit of an [APL party trick](https://aplwiki.com/wiki/Conway%27s_Game_of_Life). An excellent walk-through of the technique applied here is provided in a [Dyalog webinar](https://dyalog.tv/Webinar/?v=3FjYly2G_QI).

In [123]:
DAY18←↑'#'=¨⊃⎕NGET'data/2015/18.txt'1

In [140]:
GoL←{≢⍸⍵}⌺3 3∊¨3+0,¨,¨   ⍝ Wait, what kind of witchcraft is this...?

In [135]:
Day18p1←{+/∊GoL⍣100⊢⍵}

In [136]:
Day18p1 DAY18 ⍝ 1061

1061

For part 2, the 4 corner cells are immortal. We can simply ensure those are reset at the end of each iteration:

In [165]:
GoL2←{1@(,1 100∘.,1 100)⊢GoL ⍵}

In [166]:
Day18p2←{+/∊GoL2⍣100⊢⍵}

In [167]:
Day18p2 DAY18 ⍝ 1006

1006

### Day 19: Medicine for Rudolph
https://adventofcode.com/2015/day/19

In [169]:
DAY19←⊃⎕NGET'data/2015/19.txt'1

In [170]:
TARGET←⊃¯1↑DAY19
RULES←↓(↑' '(≠⊆⊢)¨¯2↓DAY19)[;1 3]

In [196]:
]dinput
ApplyRule←{                                 ⍝ Apply a transformation rule, returning all resulting strings
    (key val)←⍵
    ⍺∘{
        (idx len)←⍵
        (idx↑⍺),val,⍺↓⍨idx+len              ⍝ Separate the string on the match, and re-join on replacement
    }¨key ⎕S{⍵.((1↑Offsets),1↑Lengths)}⊢⍺   ⍝ Each match location
}

In [197]:
≢∪{⊃,/⊆¨⍵}⍣≡TARGET∘ApplyRule¨RULES          ⍝ 509 - count uniques in the flattened vector of produced strings

509

For part 2, we seek the smallest number of transformations to reach the target molecule starting from a single "electron": "e". To solve this we work backwards from the target molecule applying the rules in reverse until we either get to "e", or we reach a point where there is no change. 

If there is no change, we shuffle the rules order to break out of the local minimum reached in the search space. This will only result in the shortest path if there is only a single path from "e" to target, which fortunately is the case, and the convergence is super-quick.

Obviously not the most array-based solution, and most natural as a tradfn.

In [285]:
]dinput
res←molecule Day19p2 rules;current;count;prev;val;key
current←molecule
count←0
:While current≢,'e'
    prev←current
    :For val key :In rules                 ⍝ Note: we apply rules 'backwards'
        :If ~key(1∊⍷)current               ⍝ Check if rule matches
            :Continue
        :EndIf
        current←(key ⎕R val⍠'ML'1)current  ⍝ Apply single step -- as _unwinding_ transformations
        count+←1
        :If prev≡current                   ⍝ No change after applying all rules: local minimum
            current←molecule               
            count←0
            rules←(⊢⌷⍨∘⊂?⍨∘≢)rules         ⍝ Shuffle the rules ordering to break out of local minimum
        :EndIf
    :EndFor
:EndWhile
res←count

In [199]:
TARGET Day19p2 RULES ⍝ 195

195

### Day 20: Infinite Elves and Infinite Houses
https://adventofcode.com/2015/day/20

The obvious observation is that the number of elf-visits at a given house (part 1) is the sum of the divisors of the house number. There is no handy built-in for this -- the [APL Cart](https://aplcart.info/?q=sum%20of%20divisors#) suggestion of `(+/∘∪⊢∨⍳)` is too inefficient here. We can roll our own using the handy `pco` prime sieve from the [dfns](http://dfns.dyalog.com/n_pco.htm) workspace.

In [202]:
⎕CY 'dfns'       ⍝ For 'pco' -- a fast prime sieve
DAY20←29000000

In [277]:
DivSum←{×/{(¯1+⍺*⍵+1)÷⍺-1}⌿2 pco ⍵}         ⍝ Sum of positive divisors, aka σ(n)

In [278]:
Divisors←{⍵=1:,1⋄∊∘.×/{⍺∘.*0,⍳⍵}⌿2 pco ⍵}   

In [279]:
Day20p1←{(⍵÷10) {(DivSum ⍵)≥⍺:⍵⋄⍺∇⍵+1} 1}

In [280]:
Day20p1 DAY20 ⍝ 665280 -- note: takes around 15s or so to run

665280

In [283]:
]dinput
Day20p2←{
    ⍵ {sum←+/(⍵{⍺≤⍵×50}¨d)/d←Divisors ⍵
        (11×sum)≥⍺:⍵
        ⍺∇⍵+1
    } 1
}

In [284]:
Day20p2 DAY20 ⍝ 705600 -- note: takes around 30s or so to run

705600

### Day 21: RPG Simulator 20XX
https://adventofcode.com/2015/day/21

Fight-rules:

1. Must hold exactly one weapon.
2. May hold zero or one armour items.
3. May hold zero, one or two rings. If two, must be different.
4. Part 1: minimise cost whilst winning.    
5. Part 2: maximise cost whilst losing.

In [243]:
DAY21←103 9 2 ⍝ The boss-spec: Hitpoints Damage Armor

We're given the characteristics of weaponry and armor etc. Turn this into a matrix, adding an extra row of three zeros for armor, signifying that it's optional, and two such rows for the same reason in the rings section. 

In [321]:
ITEMS←8 4 0 10 5 0 25 6 0 40 7 0 74 8 0 0 0 0 13 0 1 31 0 2 53 0 3 75 0 4 102 0 5 0 0 0 0 0 0 25 1 0 50 2 0 100 3 0 20 0 1 40 0 2 80 0 3

In [324]:
(COSTS DAMAGE ARMOR)←↓⍉(3÷⍨≢ITEMS) 3⍴ITEMS

We can now construct a set of vectors representing the possible choices of 1 weapon, 0 or 1 armour, and 0, 1 or 2 rings (which must be unique).

In [295]:
⍝        <-weapon>   <---armor--->   <--------ring 1------->   <--------ring 2------->
CHOICES←,1 2 3 4 5∘.,6 7 8 9 10 11∘.,12 13 14 15 16 17 18 19∘.,12 13 14 15 16 17 18 19 
VALID←~{(⍵[3]>14)∧(⍵[3]=⍵[4])}¨CHOICES
OPTIONS←VALID/CHOICES

In [325]:
]dinput
Winner←{  ⍝ boss Winner selection
    ⍺ {
        bhp←⍺[1]-1⌈⍵[2]-⍺[3] ⋄ bhp≤0:1
        php←⍵[1]-1⌈⍺[2]-⍵[3] ⋄ php≤0:0
        (bhp@1⊢⍺)∇php@1⊢⍵
    } 100,(+/DAMAGE[⍵]),+/ARMOR[⍵]
}

In [329]:
CHARGE←{+/COSTS[⍵]}¨OPTIONS              ⍝ Total costs per kit-out in the store 
⌊/CHARGE[⍸DAY21∘Winner¨OPTIONS]          ⍝ Part 1: 121 -- least spend for a win
⌈/CHARGE[⍸~DAY21∘Winner¨OPTIONS]         ⍝ Part 2: 201 -- most spend for a loss

121

201

### Day 22: Wizard Simulator 20XX
https://adventofcode.com/2015/day/22

As in the previous question, we're given a table of data, this time describing "spells" and "effects". We massage these into a nice matrix with the following columns:

    mana-change player-hp boss-hp total-cost shield poison recharge

In [358]:
⍝ State vector fields: mana-change player-hp boss-hp total-spend shield poison recharge
⍝
⍝         <-magic missile-> <-----drain-----> <-----shield-----> <-----poison-----> <----recharge---->
⎕←MAGIC←5 7⍴¯53 0 ¯4 53 0 0 0 ¯73 2 ¯2 73 0 0 0 ¯113 0 0 113 6 0 0 ¯173 0 0 173 0 6 0 ¯229 0 0 229 0 0 5

┌→──────────────────┐
↓ ¯53 0 ¯4 53 0 0 0│
│ ¯73 2 ¯2 73 0 0 0│
│¯113 0 0 113 6 0 0│
│¯173 0 0 173 0 6 0│
│¯229 0 0 229 0 0 5│
└~──────────────────┘

This time, at each turn we must choose a spell to cast from the above five. This turns the search space into a graph, and the solution one of costed path finding.

At each point we have a state, consisting of the player and the boss with their respective "stats", plus the player's stash of "mana" and any lingering spell effects. A state transition is the casting of a spell, the application of any multi-turn spells already cast, plus the response from the boss.

In [351]:
START←500 50 71 0 0 0 0                 ⍝ State vector: mana player-hp boss-hp total-spend shield poison recharge

In [375]:
⍝ Effects
Shield←{⍵[5]>0:¯1∘+@5⊢⍵⋄⍵}              ⍝ No need to do anything apart from decrease counter
Poison←{⍵[6]>0:¯3∘+@3⊢¯1∘+@6⊢⍵⋄⍵}       ⍝ Boss: hp-←3, if active
Recharge←{⍵[7]>0:101∘+@1⊢¯1∘+@7⊢⍵⋄⍵}    ⍝ Mana+←101, if active
ApplyEffects←{Finished ⍵:⍵⋄Shield Poison Recharge ⍵}

BossTurn←{Finished ⍵:⍵⋄dp←10-7×⍵[5]>0⋄{⍵-dp}@2⊢⍵}
PlayerTurn←{⍵∘Cast¨↓MAGIC}
Finished←{(⍵[2]≤1)∨⍵[3]≤0}
StateTransition←{Finished ⍵:⍬ ⋄ ApplyEffects∘BossTurn¨PlayerTurn ApplyEffects ⍵}

In [376]:
]dinput
Cast←{               ⍝ Apply spell: state Cast spell
    ⍺[1]<-⍵[1]:⍺     ⍝ Too expensive!
    ⍵[3]<0:⍺+⍵       ⍝ Non-time limited spell; just cast it
    idx←⍸0≠¯3↑⍵
    0≠idx⌷¯3↑⍺:⍺     ⍝ Time-limited spell already in effect!
    ⍺+⍵              ⍝ Cast spell
}

We now crudely brute out the whole graph and ask questions later. This could be done much faster with a full Dijkstra or A* implementation instead.

In [377]:
]dinput
BFS←{                ⍝ Naive breadth-first search, to the bottom. Accumulate visited states as ⍺
    ⍬ {                          
        0=≢⍵:⍺                  
        current←⊃1↑⍵ ⋄ queue←1↓⍵
        neighbours←StateTransition current
        (⍺,⊂current)∇queue∪neighbours~⍺
    } ⍵                          
}

In [378]:
]dinput
Day22p1←{
    data←⍺ BFS ⍵
    bhp←(↑data)[;3]       ⍝ Boss hitpoints
    ⌊/(↑(0>bhp)/data)[;4] ⍝ Pick winning rows, and find smallest total cost
}

In [367]:
Day22p1 ⊂START ⍝ 1824 -- note, takes a minute or two to run.

1824

### Day 23: Opening the Turing Lock
https://adventofcode.com/2015/day/23

In [444]:
DAY23←⊃⎕NGET'data/2015/23.txt'1

In [445]:
RegexGroups←{⍺⎕S{⍵.(1↓Lengths↑¨Offsets↓¨⊂Block)} ⊢ ⍵}
Parse←{0<≢'^[+-]?\d+$'⎕S'&'⊢⍵:⍎⍵⋄⍵≡,'a':1⋄⍵≡,'b':2⋄⍵}

In [446]:
data←{⊃,/⊆¨⍵}⍣≡'([a-z]{3}) ([+ab0-9-]+),?\s*([+0-9]*)'∘RegexGroups¨DAY23

In [447]:
MEM←(3÷⍨≢data) 3⍴Parse¨data

In [458]:
]dinput
Day23←{
    ip←3⌷⍺ 
    ip>≢⍵:⍺
    instr←ip⌷⍵⋄op←⊃instr                                     ⍝ Fetch instruction at ip, opcode is first element
    op≡'hlf':(⍺ {r←⊃⍵⋄1∘+@3⊢(2÷⍨r⌷⍺)@r⊢⍺} 1↓instr)∇⍵         ⍝ Halve reg content, add 1 to ip
    op≡'jio':(⍺ {(r d)←⍵⋄1=r⌷⍺:d∘+@3⊢⍺⋄1∘+@3⊢⍺} 1↓instr)∇⍵   ⍝ If reg content is 1, add delta to ip, else add 1
    op≡'inc':(⍺ {r←⊃⍵⋄1∘+@3⊢(1+⍨r⌷⍺)@r⊢⍺} 1↓instr)∇⍵         ⍝ Add 1 to reg content, add 1 to ip
    op≡'tpl':(⍺ {r←⊃⍵⋄1∘+@3⊢(3×⍨r⌷⍺)@r⊢⍺} 1↓instr)∇⍵         ⍝ 3 × reg content, add 1 to ip
    op≡'jmp':(⍺ {d←⊃⍵⋄d∘+@3⊢⍺} 1↓instr)∇⍵                    ⍝ Add delta to ip
    op≡'jie':(⍺ {(r d)←⍵⋄0=2|r⌷⍺:d∘+@3⊢⍺⋄1∘+@3⊢⍺} 1↓instr)∇⍵ ⍝ If reg content is even, add delta to ip, else add 1
}

In [459]:
2⌷0 0 1 Day23 MEM ⍝ 255

255

In [457]:
2⌷1 0 1 Day23 MEM ⍝ 334

334

### Day 24: It Hangs in the Balance
https://adventofcode.com/2015/day/24

Data is a list of very prime-looking numbers. Task is to create three groups, summing to the same number, with the extra constraint that the first group should minimise the number of items.

In [464]:
'cmat' 'big'⎕CY'dfns'
DAY24←1 3 5 11 13 17 19 23 29 31 37 41 43 47 53 59 67 71 73 79 83 89 97 101 103 107 109 113

In [465]:
3÷⍨+/DAY24

508

By applying part guesswork and part brute force, we will start at the minimum number of items required, try all combinations, and if none is found, add another item such that the sum is 508 (data sum divided by 3). In other words, we only search for the first group, and "hope" that the rest works out. The lower bound of the number of items is 5, as 103 + 107 + 109 + 113 < 508, but adding the fifth largest item takes us above.

In [466]:
]dinput
TrySelection←{
    numbers←⍺
    (elements val)←⍵
    Best←{⍺≡⍬:⍵⋄(×/⍺)<×/⍵:⍺⋄⍵}             ⍝ Compare products of elements
    ⍬{
        0=≢⍵:⍺
        weights←numbers[1↑⍵]               ⍝ Selected elements
        val=+/weights:(⍺ Best weights)∇1↓⍵ ⍝ Correct sum? If so, any better than previous best?
        ⍺∇1↓⍵                              ⍝ Keep going
    }↑elements cmat ≢numbers               ⍝ All possible selection of 'elements' numbers
}

In [470]:
Day24←{val←⍺÷⍨+/⍵⋄⍵ {v←⍺ TrySelection ⍵ val⋄v≢⍬:v⋄∇⍵+1} 5} ⍝ Starting from 5 elements, increase until solution

In [468]:
⊃×big/3 Day24 DAY24 ⍝ Part 1: 10439961859 -- need to use big-number × from dfns

┌→──────────┐
│10439961859│
└───────────┘

In [471]:
⊃×big/4 Day24 DAY24 ⍝ Part 2: 72050269

┌→───────┐
│72050269│
└────────┘

### Day 25: Let It Snow
https://adventofcode.com/2015/day/25

In [472]:
DAY25←2947 3029